node/network/gossip-support/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This subsystem is responsible for keeping track of session changes
+//! and issuing a connection request to the validators relevant to
+//! the gossiping subsystems on every new session.
+
+use futures::FutureExt as _;
+use polkadot_node_subsystem::{
+	messages::{
+		GossipSupportMessage,
+	},
+	ActiveLeavesUpdate, FromOverseer, OverseerSignal,
+	Subsystem, SpawnedSubsystem, SubsystemContext,
+};
+use polkadot_node_subsystem_util::{
+	validator_discovery::{ConnectionRequest, self},
+	self as util,
+};
+use polkadot_primitives::v1::{
+	Hash, ValidatorId, SessionIndex,
+};
+use polkadot_node_network_protocol::peer_set::PeerSet;
+
+const LOG_TARGET: &str = "gossip_support";
+
+/// The Gossip Support subsystem.
+pub struct GossipSupport {}
+
+#[derive(Default)]
+struct State {
+	last_session_index: Option<SessionIndex>,
+	/// when we overwrite this, it automatically drops the previous request
+	last_connection_request: Option<ConnectionRequest>,
+}
+
+impl GossipSupport {
+	/// Create a new instance of the [`GossipSupport`] subsystem.
+	pub fn new() -> Self {
+		Self {}
+	}
+
+	#[tracing::instrument(skip(self, ctx), fields(subsystem = LOG_TARGET))]
+	async fn run<Context>(self, mut ctx: Context)
+	where
+		Context: SubsystemContext<Message = GossipSupportMessage>,
+	{
+		let mut state = State::default();
+		loop {
+			let message = match ctx.recv().await {
+				Ok(message) => message,
+				Err(e) => {
+					tracing::debug!(
+						target: LOG_TARGET,
+						err = ?e,
+						"Failed to receive a message from Overseer, exiting"
+					);
+					return;
+				},
+			};
+			match message {
+				FromOverseer::Communication { .. } => {},
+				FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate {
+					activated,
+					..
+				})) => {
+					tracing::trace!(target: LOG_TARGET, "active leaves signal");
+
+					let leaves = activated.into_iter().map(|(h, _)| h);
+					if let Err(e) = state.handle_active_leaves(&mut ctx, leaves).await {
+						tracing::debug!(target: LOG_TARGET, "Error {}", e);
+					}
+				}
+				FromOverseer::Signal(OverseerSignal::BlockFinalized(_hash, _number)) => {},
+				FromOverseer::Signal(OverseerSignal::Conclude) => {
+					return;
+				}
+			}
+		}
+	}
+}
+
+async fn determine_relevant_validators(
+	ctx: &mut impl SubsystemContext,
+	relay_parent: Hash,
+	_session: SessionIndex,
+) -> Result<Vec<ValidatorId>, util::Error> {
+	let validators = util::request_validators_ctx(relay_parent, ctx).await?.await??;
```

Comment:
Please file a follow-on for avoiding connecting to all validators. Instead some randomly selected subset of size sqrt(n) with, say min=25 would help us with scaling. our initial scaling goals are 200, and then 1000, and then 10000.

Reply:
resolved in 404f37c

Reply:
In this [document](https://research.web3.foundation/en/latest/polkadot/networking/3-avail-valid.html) under Topology, an interesting scheme is described, which explains how validators of different groups should decide to whom they should be connected to. It is meant to provide good connectivity and as evenly distributed connections as possible.

Reply:
The problem with complete random picking of nodes is twofold.

1. We are not guaranteed to have a fully connected graph.
2. Connections will be likely distributed in a rather [unbalanced way](https://theartofmachinery.com/2020/01/27/systems_programming_probability.html).

I guess it is good enough for now. I will stay in touch with research on how we should proceed with this longer term, filing issues as necessary.

Reply:
@infinity0